### PR TITLE
[DSpace-CRIS] Fixed Recipients Email Feature

### PR DESCRIPTION
**Pull request by vins01-4science**
## Description
This PR introduces a new feature that will allow to send all the emails to a given list of emails ( `catchAllRecipients` ).
- When `mail.server.catchAll.enabled=true` and `mail.server.catchAll.recipient` is configured:  
  - * All emails are sent to catchAll recipients instead of original recipients
  - * Original recipients are logged and appended to email body for reference
  - * Useful for development and test environments

## Instructions for Reviewers
Try to configure the following properties:
-  `mail.server.catchAll.recipient = youremail`
-  `mail.server.catchAll.enabled=true`
Then try to sent an email, you should verify that the email will be sent to that specified email address.

List of changes in this PR:
- Added support for catchAll.recipient in Email.java
- Added catchAll configurations in dspace.cfg and local.cfg.EXAMPLE
- Added Arrays import and ccAddresses field to Email.java
- Implemented fixedRecipient logic in build() method
- When `mail.server.disabled=true` and `mail.server.catchAll.enabled=true` is configured:  
  - * All emails are sent to catchAll recipients instead of original recipients
  - * Original recipients are logged and appended to email body for reference
  - * Useful for development and test environments

## Checklist

- [x] My PR is **created against the `main` branch** of code (unless it is a backport or is fixing an issue specific to an older branch).
- [x] My PR is **small in size** (e.g. less than 1,000 lines of code, not including comments & integration tests). Exceptions may be made if previously agreed upon.
- [x] My PR **passes Checkstyle** validation based on the [Code Style Guide](https://wiki.lyrasis.org/display/DSPACE/Code+Style+Guide).
- [x] My PR **includes Javadoc** for _all new (or modified) public methods and classes_. It also includes Javadoc for large or complex private methods.
- [x] My PR **passes all tests and includes new/updated Unit or Integration Tests** based on the [Code Testing Guide](https://wiki.lyrasis.org/display/DSPACE/Code+Testing+Guide).
- [x] My PR **includes details on how to test it**. I've provided clear instructions to reviewers on how to successfully test this fix or feature.
- [x] If my PR includes new libraries/dependencies (in any `pom.xml`), I've made sure their licenses align with the [DSpace BSD License](https://github.com/DSpace/DSpace/blob/main/LICENSE) based on the [Licensing of Contributions](https://wiki.lyrasis.org/display/DSPACE/Code+Contribution+Guidelines#CodeContributionGuidelines-LicensingofContributions) documentation.
- [x] If my PR modifies REST API endpoints, I've opened a separate [REST Contract](https://github.com/DSpace/RestContract/blob/main/README.md) PR related to this change.
- [x] If my PR includes new configurations, I've provided basic technical documentation in the PR itself.
- [x] If my PR fixes an issue ticket, I've [linked them together](https://docs.github.com/en/issues/tracking-your-work-with-issues/linking-a-pull-request-to-an-issue).

### CODE REVIEW

File: dspace/config/dspace.cfg [hunk 1]
```diff
@@ -205,6 +205,8 @@ mail.charset = UTF-8
 # It will instead log the subject of the email which should have been sent
 # This is especially useful for development and test environments where production data is used when testing functionality.
 #mail.server.disabled = false
+# If the mail.server.fixedRecipient is not empty then only in the case mail.server.disabled = true all email will be sent to this email address.
+#mail.server.fixedRecipient =
```
Comment (paulo-graca) [inline dspace/config/dspace.cfg]: I would like to suggest one additional configuration. I think it would clarify the purpose of this configuration.

from:
  `mail.server.fixedRecipient = my@email.com`
to something like:
  `catchAll.recipient = my@email.com`
(it would be better aligned with what we already have, like: `feedback.recipient` or `alert.recipient`)

and also add a setting to control it's usage, something like:
  `mail.catchAllAccount.enabled = false`

File: dspace/config/local.cfg.EXAMPLE [hunk 2]
```diff
@@ -134,6 +134,14 @@ db.password = dspace
 # Recipient for new user registration emails (defaults to unspecified)
 #registration.notify =
 
+# An option is added to disable the mailserver. By default, this property is set to false
+# By setting mail.server.disabled = true, DSpace will not send out emails.
+# It will instead log the subject of the email which should have been sent
+# This is especially useful for development and test environments where production data is used when testing functionality.
+#mail.server.disabled = false
+# If the mail.server.fixedRecipient is not empty then only in the case mail.server.disabled = true all email will be sent to this email address.
+#mail.server.fixedRecipient =
```
Comment (paulo-graca) [inline dspace/config/local.cfg.EXAMPLE]: if you consider to change the configuration naming, then, this documentation should be also updated.

File: dspace-api/src/main/java/org/dspace/core/Email.java [hunk 3]
```diff
@@ -607,7 +664,7 @@ public static void main(String[] args) {
             System.out.println(" - To: " + to);
             System.out.println(" - Subject: " + subject);
             System.out.println(" - Server: " + server);
-            boolean disabled = config.getBooleanProperty("mail.server.disabled", false);
+            boolean disabled = isMailServerDisabled(config);
```
Comment (paulo-graca) [inline dspace-api/src/main/java/org/dspace/core/Email.java:661]: `mail.server.disabled` should prevail. The expected behavior should be, If true no emails are to be sent. It's what it's expected to do.

File: dspace-api/src/main/java/org/dspace/core/Email.java [hunk 4]
```diff
@@ -348,16 +355,24 @@ public void reset() {
...
     }
 
+    /**
+     * Checks if the generated email is valid, and can be sent.
+     * @return boolean
+     */
+    private static boolean isValidEmail() {
+        ConfigurationService config = DSpaceServicesFactory.getInstance().getConfigurationService();
+        boolean disabled = isMailServerDisabled(config);
+        String[] fixedRecipient = config.getArrayProperty("mail.server.fixedRecipient", new String[] { });
```
Comment (paulo-graca) [inline dspace-api/src/main/java/org/dspace/core/Email.java]: it's a matter of preference. I prefer `catchAllAccount` or `catchAllRecipient` concept instead of a `fixedRecipient`. To me, I assume that a `catchAllAccount` is to where all emails are delivered.

File: dspace-api/src/main/java/org/dspace/core/Email.java [hunk 5]
```diff
@@ -396,11 +411,9 @@ void build()
         // Create message
         message = new MimeMessage(session);
 
-        // Set the recipients of the message
-        for (String recipient : recipients) {
-            message.addRecipient(Message.RecipientType.TO,
-                    new InternetAddress(recipient));
-        }
+        // Get the mail configuration properties for fixedRecipient
+        String[] fixedRecipients = getFixedRecipients(config);
```
Comment (paulo-graca) [inline dspace-api/src/main/java/org/dspace/core/Email.java]: I would prefer to keep the existing code of doing a cycle for any configured recipients. Just adding a validation for `mail.catchAllAccount.enabled = true` to use the catchAll alternative instead.

**Review by paulo-graca — COMMENTED** [review]
Thank you @vins01-4science!

I believe I understand the purpose or use case — it applies when you have a repository where you want to keep the existing email configuration (though others might wonder why not simply change the email settings in your `local.cfg`), but still need all outgoing mail to be redirected to a specific email account. This is mainly useful for testing environments.

In my opinion, the current approach of hacking the `mail.server.disabled` property is not ideal, as the installer will expect the standard behavior you’re overriding. Instead, I would suggest introducing a separate configuration to control this mechanism, and I’d also propose a different naming convention. Personally, I prefer the term catch-all instead of fixed.

(For context, we’re implementing the same kind of feature at the mail server level across all our test environments.)

File: dspace-api/src/test/java/org/dspace/core/EmailTest.java [hunk 6]
```diff
@@ -65,4 +68,157 @@ public void testNotNullParameter()
...
+        Email email = new Email();
+        email.setContent("fixed recipient test",
+                "This is a test email that should go to catchAll recipient.");
+        email.addRecipient("original@example.com");
+        email.build();
+
+        MimeMessage message = email.message;
+        InternetAddress[] recipients = (InternetAddress[]) message.getRecipients(MimeMessage.RecipientType.TO);
+
+        assertThat("Email should be sent to catchAll recipient", recipients.length, is(1));
```
Comment (mwoodiupui) [inline dspace-api/src/test/java/org/dspace/core/EmailTest.java]: "Exactly one 'To:' recipient expected."

File: dspace-api/src/test/java/org/dspace/core/EmailTest.java [hunk 7]
```diff
@@ -65,4 +68,157 @@ public void testNotNullParameter()
         assertThat("Null message parameter should be transformed to empty",
                 message, containsString(testParam));
     }
+
+    @Test
+    public void testCatchAllRecipientWithDisabledServer()
+            throws MessagingException, IOException {
+        config.setProperty("mail.server.catchAll.enabled", "true");
+        config.setProperty("mail.server.catchAll.recipient", "fixed@example.com");
```
Comment (mwoodiupui) [inline dspace-api/src/test/java/org/dspace/core/EmailTest.java:78]: mail.server.disabled is not set true as the test's name implies.

File: dspace-api/src/test/java/org/dspace/core/EmailTest.java [hunk 8]
```diff
@@ -65,4 +68,157 @@ public void testNotNullParameter()
...
+        Email email = new Email();
+        email.setContent("normal sending test",
+                "This is a test email that should go to original recipient.");
+        email.addRecipient("original@example.com");
+        email.build();
+
+        MimeMessage message = email.message;
+        InternetAddress[] recipients = (InternetAddress[]) message.getRecipients(MimeMessage.RecipientType.TO);
+
+        assertThat("Email should be sent to original recipient when server is enabled",
```
Comment (mwoodiupui) [inline dspace-api/src/test/java/org/dspace/core/EmailTest.java]: "Message should have exactly one recipient."

File: dspace-api/src/test/java/org/dspace/core/EmailTest.java [hunk 9]
```diff
@@ -65,4 +68,157 @@ public void testNotNullParameter()
...
+        email.addRecipient("original@example.com");
+        // Note: In the current implementation, CC addresses need to be added via the message directly
+        // For this test, we'll verify the structure is ready for CC handling
+        email.build();
+
+        MimeMessage message = email.message;
+        InternetAddress[] recipients = (InternetAddress[]) message.getRecipients(MimeMessage.RecipientType.TO);
+        InternetAddress[] ccRecipients = (InternetAddress[]) message.getRecipients(MimeMessage.RecipientType.CC);
+
+        assertThat("Email should be sent to catchAll recipient", recipients.length, is(1));
```
Comment (mwoodiupui) [inline dspace-api/src/test/java/org/dspace/core/EmailTest.java]: "Message should have exactly one recipient."

File: dspace-api/src/test/java/org/dspace/core/EmailTest.java [hunk 10]
```diff
@@ -65,4 +68,157 @@ public void testNotNullParameter()
...
+        assertThat("CatchAll recipient should receive the email", recipients[0].getAddress(), is("fixed@example.com"));
+        assertThat("CC field should be null for catchAll recipient", ccRecipients, is((InternetAddress[]) null));
```
Comment (mwoodiupui) [inline dspace-api/src/test/java/org/dspace/core/EmailTest.java:198]: Possibly should test for a zero-element array as well.  The `Message` documentation doesn't say what happens when `setRecipients` is given a zero-length string or an empty array.

File: dspace/config/dspace.cfg [hunk 11]
```diff
@@ -205,6 +205,11 @@ mail.charset = UTF-8
 # It will instead log the subject of the email which should have been sent
 # This is especially useful for development and test environments where production data is used when testing functionality.
 #mail.server.disabled = false
+# Enables the catchAll system, more details downer.
```
Comment (mwoodiupui) [inline dspace/config/dspace.cfg]: ", more details below" or simply omit the phrase.

File: dspace/config/local.cfg.EXAMPLE [hunk 12]
```diff
@@ -134,6 +134,17 @@ db.password = dspace
 # Recipient for new user registration emails (defaults to unspecified)
 #registration.notify =
 
+# An option is added to disable the mailserver. By default, this property is set to false
+# By setting mail.server.disabled = true, DSpace will not send out emails.
+# It will instead log the subject of the email which should have been sent
+# This is especially useful for development and test environments where production data is used when testing functionality.
+#mail.server.disabled = false
+# Enables the catchAll system, more details downer.
```
Comment (mwoodiupui) [inline dspace/config/local.cfg.EXAMPLE]: "below" or omit the phrase.

**Review by mwoodiupui — CHANGES_REQUESTED** [review]
This looks well overall.  I would suggest that each test set `mail.server.disabled` to the desired state.  Some assertions could better describe what they actually test.

File: dspace-api/src/test/java/org/dspace/core/EmailTest.java (hunk 10 above)
Comment (vins01-4science) [reply dspace-api/src/test/java/org/dspace/core/EmailTest.java:198]: I think it's already there: `testDisabledServerWithoutCatchAllRecipient`.
Created another test for the null case.

**Comment by vins01-4science** [comment]
> This looks well overall. I would suggest that each test set `mail.server.disabled` to the desired state. Some assertions could better describe what they actually test.

Changes applied. Let me know if you find something else!

File: dspace-api/src/test/java/org/dspace/core/EmailTest.java [hunk 13]
```diff
@@ -65,4 +69,189 @@ public void testNotNullParameter()
...
+        assertThat("First catchAll recipient should receive the email", recipients[0].getAddress(),
+                   is("fixed1@example.com"));
+        assertThat("Second catchAll recipient should receive the email", recipients[1].getAddress(),
+                   is("fixed2@example.com"));
+    }
+
+    @Test
+    public void testDisabledServerWithoutCatchAllRecipient()
+            throws MessagingException, IOException {
+        config.setProperty("mail.server.disabled", "false");
```
Comment (mwoodiupui) [inline dspace-api/src/test/java/org/dspace/core/EmailTest.java:126]: Name is testDisabledServer... but server is not disabled.

File: dspace-api/src/test/java/org/dspace/core/EmailTest.java [hunk 14]
```diff
@@ -65,4 +69,189 @@ public void testNotNullParameter()
...
+
+        assertThat("Message should be created", message, is(not((MimeMessage) null)));
+        assertThat("Email should be sent to original recipient when no catchAll recipient configured",
+                recipients.length, is(1));
+        assertThat("Original recipient should receive the email", recipients[0].getAddress(),
+                   is("original@example.com"));
+
+        // Check that original recipient is NOT included in the body (normal behavior)
+        String messageBody = message.getContent().toString();
+        assertThat("Original recipient should NOT be included in email body for normal disabled behavior",
```
Comment (mwoodiupui) [inline dspace-api/src/test/java/org/dspace/core/EmailTest.java]: Confusing assertion label:  the test is for the REAL RECIPIENT header, not for the original recipient.  The comment above is also not quite correct.

File: dspace-api/src/test/java/org/dspace/core/EmailTest.java [hunk 15]
```diff
@@ -65,4 +69,189 @@ public void testNotNullParameter()
...
+        MimeMessage message = email.message;
+        InternetAddress[] recipients = (InternetAddress[]) message.getRecipients(MimeMessage.RecipientType.TO);
+
+        assertThat("Message should have exactly one recipient.", recipients.length, is(1));
+        assertThat("Original recipient should receive the email",
+                   recipients[0].getAddress(), is("original@example.com"));
+
+        // Check that original recipient is NOT included in the body (normal behavior)
+        String messageBody = message.getContent().toString();
+        assertThat("Original recipient should NOT be included in email body for normal sending",
```
Comment (mwoodiupui) [inline dspace-api/src/test/java/org/dspace/core/EmailTest.java]: Confusing comment and assertion text, as above.

File: dspace-api/src/test/java/org/dspace/core/EmailTest.java [hunk 16]
```diff
@@ -65,4 +69,189 @@ public void testNotNullParameter()
...
+        MimeMessage message = email.message;
+        InternetAddress[] recipients = (InternetAddress[]) message.getRecipients(MimeMessage.RecipientType.TO);
+        InternetAddress[] ccRecipients = (InternetAddress[]) message.getRecipients(MimeMessage.RecipientType.CC);
+
+        assertThat("Message should have exactly one recipient.", recipients.length, is(1));
+        assertThat("Original recipient should receive the email", recipients[0].getAddress(), is("original@example.com"));
+        assertThat("CC field should be null for catchAll recipient", ccRecipients, Matchers.nullValue());
+
+        assertThat(
+            "Original recipient should NOT be included in email body for normal disabled behavior",
```
Comment (mwoodiupui) [inline dspace-api/src/test/java/org/dspace/core/EmailTest.java]: Text does not describe the test.

**Review by mwoodiupui — CHANGES_REQUESTED** [review]
There are still a few places where the assertion text is misleading.  I'm sorry if I missed these earlier.

**Review by paulo-graca — APPROVED** [review]
Thank you @vins01-4science for addressing naming! I'm also following @mwoodiupui comments, but for me it's much clear now.

**Comment by vins01-4science** [comment]
> There are still a few places where the assertion text is misleading. I'm sorry if I missed these earlier.

The requested changes have been applied. Do you have any other feedback @mwoodiupui on this work ?

**Review by mwoodiupui — APPROVED** [review]

**Comment by vins01-4science** [comment]
@tdonohue - I think we can proceed with the merge of this feature.

Thank you @mwoodiupui , @paulo-graca , @tdonohue

File: dspace/config/local.cfg.EXAMPLE [hunk 17]
```diff
@@ -134,6 +134,17 @@ db.password = dspace
...
+# An option is added to disable the mailserver. By default, this property is set to false
+# By setting mail.server.disabled = true, DSpace will not send out emails.
+# It will instead log the subject of the email which should have been sent
+# This is especially useful for development and test environments where production data is used when testing functionality.
+#mail.server.disabled = false
+# Enables the catchAll system
+#mail.server.catchAll.enabled = true
+# If the mail.server.catchAll.recipient is set and has been enabled with mail.server.catchAll.enabled all the emails will
+# be sent to these recipients ( comma separated list ), instead of original ones.
+#mail.server.catchAll.recipient =
```
Comment (tdonohue) [inline dspace/config/local.cfg.EXAMPLE]: @vins01-4science : Could we remove these settings from `local.cfg.EXAMPLE`?  I feel the `local.cfg.EXAMPLE` file should only contain settings which are **often modified in production**, as it's a "starting point" for creating your own `local.cfg`.

These settings are _unlikely to be used in Production scenarios_ (as they are more for test/development), so I feel they should only be included in `dspace.cfg`.

**Review by tdonohue — COMMENTED** [review]
@vins01-4science : Thanks, this looks good to me overall.  I also tested it and found it works as described.  That said, I have a *minor* request... could you remove these configs from the `local.cfg.EXAMPLE`?  That file is meant to only contain settings which are frequently changed in Production scenarios, and these new settings are not really meant for production.

Once that minor change is made, I think this is ready to be merged.

File: dspace/config/local.cfg.EXAMPLE (hunk 17 above)
Comment (vins01-4science) [reply dspace/config/local.cfg.EXAMPLE]: Thank you for the suggestion! That makes sense.
 I will move these settings from local.cfg.EXAMPLE to dspace.cfg as recommended.

**Review by tdonohue — APPROVED** [review]
👍 Thanks @vins01-4science !  This looks good to me now.  Merging immediately.

However, please note, I'm going to flag this as "Needs Documentation" just because it needs docs added to https://wiki.lyrasis.org/display/DSDOC10x regarding the new configurations.  Likely they could be added to the [Configuration Reference](https://wiki.lyrasis.org/display/DSDOC10x/Configuration+Reference) alongside the current Email settings.

**Comment by GraziaQuercia** [comment]
documentation ready https://wiki.lyrasis.org/display/DSDOC10x/Sending+emails+to+fixed+recipients
